static/metrics/agent.go

```diff
@@ -227,9 +230,16 @@ func (a *Agent) ApplyConfig(cfg Config) error {
 	a.mut.Lock()
 	defer a.mut.Unlock()
 
-	if util.CompareYAML(a.cfg, cfg) {
+	newConfigByteArr, err := yaml.Marshal(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to marshal new config: %w", err)
+	}
+	newConfig := string(newConfigByteArr[:])
+	if newConfig == a.previousConfig {
+		level.Debug(a.logger).Log("msg", "not recreating metrics instance because config hasn't changed")
 		return nil
 	}
+	a.previousConfig = newConfig
```

Comment:
I think there might be a subtle bug in both the old code and the new code around secrets.

All secrets are marshaled as a generic string when calling `yaml.Marshal`. I suspect if the only thing that changes between two reloads is a password field, static mode will incorrectly think nothing has changed and skip the update. 

That's the reason we have `util.CompareYAMLWithHook`; to allow us to bypass the generic string marshaling for secrets and fully compare two YAMLs. 

Can you double check this?

Reply:
I tested the hook with a metrics config, and unfortunately it didn't work. E.g. this test doesn't pass:

```
func TestMarshalWithHook2(t *testing.T) {
	cfg := metrics.Config{
		Configs: []instance.Config{
			{
				ScrapeConfigs: []*prom_config.ScrapeConfig{
					{
						HTTPClientConfig: config_util.HTTPClientConfig{
							BasicAuth: &config_util.BasicAuth{
								Username: "user",
								Password: config_util.Secret("pass"),
							},
						},
					},
				},
			},
		},
	}

	expectedYaml := `configs:
- scrape_configs:
  - job_name: ""
    honor_timestamps: false
    track_timestamps_staleness: false
    basic_auth:
      username: user
      password: pass
    follow_redirects: false
    enable_http2: false
`

	actualYaml, err := util.MarshalWithHook(cfg, util.NoScrubbedSecretsHook)
	require.NoError(t, err)

	require.Equal(t, expectedYaml, string(actualYaml))
}
```

The reason is the call to `MarshalConfig(&c, true)` in `/static/metrics/instance/instance.go`. I can't think of a good way to propagate the preference to not marshal to `<secret>`. Global variables won't work, since there could be more than one thread marshalling at any given time.

Do you mind if we merge the PR as it is, and I could raise an issue for us to tackle the problem with secrets separately? In the end it might be best to compare the configs without using yaml.

Reply:
Since the bug was already present, sure, I don't mind us merging this as-is.

Reply:
Thanks! I opened the [issue](https://github.com/grafana/agent/issues/6978).